linkerd/http-route/src/lib.rs

```diff
@@ -0,0 +1,103 @@
+//! An HTTP route matching library for Linkerd to support the HTTPRoute (and
+//! GRPCRoute) Kubernetes Gateway API types.
+
+#![deny(rust_2018_idioms, clippy::disallowed_methods, clippy::disallowed_types)]
+#![forbid(unsafe_code)]
+
+use tracing::trace;
+
+pub mod grpc;
+pub mod http;
+
+// Matchers used by both HTTP and gRPC routes.
+pub use self::http::{HostMatch, MatchHeader, MatchHost};
+
+/// Groups routing rules under a common set of hostnames.
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct Route<M, P> {
+    /// A list of hostnames that this route applies to, to be matched against,
+    /// e.g., the HTTP `host`.
+    ///
+    /// If at least one match is specified, any match may apply to the request
+    /// for rules to applied. When no host matches are present, all hosts match.
+    pub hosts: Vec<http::MatchHost>,
+
+    /// Must not be empty.
+    pub rules: Vec<Rule<M, P>>,
+}
+
+/// Policies for a given set of route matches.
+#[derive(Clone, Debug, Default, Hash, PartialEq, Eq)]
+pub struct Rule<M, P> {
+    /// A list of request matchers, *any* of which may apply.
+    ///
+    /// The "best" match is used when comparing rules.
+    pub matches: Vec<M>,
+
+    /// The policy to apply to requests matched by this rule.
+    pub policy: P,
+}
+
+/// Summarizes a matched route so that route matches may be compared/ordered. A
+/// greater matches is preferred over a lesser match.
+#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct RouteMatch<T> {
+    host: Option<http::HostMatch>,
+    route: T,
+}
+
+/// A strategy for matching a request to a route.
+pub trait Match {
+    type Summary: Default + Ord;
+
+    fn r#match<B>(&self, req: &::http::Request<B>) -> Option<Self::Summary>;
```

Comment:
this is maybe silly, but there is a part of me that feels like maybe we should just rename this to `find_match` or something like that rather than having to type `r#match`? :woman_shrugging: